### ISSUE TITLE
⟦Expand Stroke⟧ Stroking certain path results in weird degenerate spline

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3694]
# ☞ See below for test case.

![2019-05-09-160009_2560x1401_scrot](https://user-images.githubusercontent.com/838783/57437002-6e5e7380-7273-11e9-9c14-15db74f41e00.png)
![2019-05-09-160619_2204x1895_scrot](https://user-images.githubusercontent.com/838783/57437395-47ed0800-7274-11e9-8854-1d027bf0c897.png)

## stdout
```
Internal Error (overlap): The curve is too short.
Internal Error (overlap): The curve is too short.
Internal Error (overlap): monotonic is both needed and unneeded (451.785,167.938)->(361.457,140.992). y=140.992 (prev=323.091)
Internal Error (overlap): Winding number did not return to 0 when y=140.992
Internal Error (overlap): monotonic is both needed and unneeded (451.785,167.938)->(361.457,140.992). y=140.992 (prev=323.091)
Internal Error (overlap): Winding number did not return to 0 when y=140.992
Internal Error (overlap): monotonic is both needed and unneeded (451.785,167.938)->(361.457,140.992). y=140.992 (prev=323.091)
Internal Error (overlap): Winding number did not return to 0 when y=140.992
```

## Background
I've found another Expand Stroke bug. As an avid fan of Spiro and Expand Stroke, I of course constantly suffer from its bugginess while working on my fonts, and recently found a workaround to its most annoying bug in https://github.com/fontforge/fontforge/issues/3264#issuecomment-489408743, but this most recent bug seems to me something totally new and undiscovered.

# Glyph `qof` (ק)

Here is the spline that causes it:

<img src="https://user-images.githubusercontent.com/838783/57437328-1f650e00-7274-11e9-9d65-fef1f5f329b4.png" width="200px">

When simplified (_Element&rarr;Simplify&rarr;Simplify_), it no longer happens:

<img src="https://user-images.githubusercontent.com/838783/57437768-0ad54580-7275-11e9-8229-92c6ee5f7608.png" width="200px">

I'm not claiming that this ק is exceedingly beautiful, but certainly it's not so ugly that it should cause FontForge to behave this way, right? :wink:

Well, the spiral seems to be the main problem. I found other weird things:

# Glyph `poof`

If I remove a lot of the points and just leave the little Spiro, when I try to expand it it disappears (glyph `poof`):

<img src="https://user-images.githubusercontent.com/838783/57438552-c21e8c00-7276-11e9-9abd-79f7fe1fd53f.png" width="200px">

(no stdout)

# Glyph `degen2`

If I leave some of the more points of the spiro, I get a very weird looking degenerate case.

<img src="https://user-images.githubusercontent.com/838783/57438742-30fbe500-7277-11e9-819b-2d0d258ffc5f.png" width="200px">
<img src="https://user-images.githubusercontent.com/838783/57438930-86d08d00-7277-11e9-8bdf-312262daba89.png" width="200px">

All the other ones are degenerate and not correct, but this is the only one that gives a warning: 
`Warning: Unreasonably big splines. They will be ignored.`

I played around with it some more, but all cases seem to fit into `qof` (minor degeneracy), `degen2` (major degeneracy), or `poof` (total disappearance).

# Necessary settings

Pen Type: Circular;
Main & Minor Stroke Width: 50;
Pen Angle: 45;
**Line Cap: Round;**&dagger;
Line Join: Round.
Remove neither internal nor external contours.

&dagger; As far as I know only this setting is non-default, but for the convenience of bug hunters I provide all my settings.

# Test case file

**Comment by frank-trampe** [comment]
I used gdb to skip overlap removal in the stroking process. The initial stroking is mostly correct, but it has one wild control point that somehow makes the roof come down during overlap removal. In fact, not even the display rendering code can deal with it. The curve goes a different direction depending upon the zoom window.

![Screenshot from 2019-06-20 16-25-36](https://user-images.githubusercontent.com/2512464/59882639-66b4e300-9378-11e9-9c99-71223a1387a5.png)
![Screenshot from 2019-06-20 16-26-02](https://user-images.githubusercontent.com/2512464/59882640-674d7980-9378-11e9-9c38-5a0d2dcf0307.png)
[qof-dist-stroke-only-1.sfd.gz](https://github.com/fontforge/fontforge/files/3312179/qof-dist-stroke-only-1.sfd.gz)

**Comment by frank-trampe** [comment]
I am inclined to deal with this specific problem by discarding any wild control points in the primary stroking result (prior to overlap removal) and have pushed some code to do that. It leaves unresolved the issues in the stroking that cause that to happen and the inability of the overlap remover to deal with such points.

**Comment by ctrlcctrlv** [comment]
This can be closed, it was fixed by the merge of #3747.
